components/collections/src/codepointinvlist/cpinvlist.rs

```diff
@@ -74,17 +115,31 @@ impl databake::Bake for CodePointInversionList<'_> {
     }
 }
 
-// Note: serde(flatten) currently does not promote a struct field of type Vec
-// to replace the struct when serializing. The error message from the default
-// serialization is: "can only flatten structs and maps (got a sequence)".
-
 #[cfg(feature = "serde")]
 impl<'data> serde::Serialize for CodePointInversionList<'data> {
     fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
     where
         S: serde::Serializer,
     {
-        self.inv_list.serialize(serializer)
+        if serializer.is_human_readable() {
+            use serde::ser::SerializeSeq;
+            let mut seq = serializer.serialize_seq(Some(self.inv_list.len() / 2))?;
+            for range in self.iter_ranges() {
+                let start = unsafe { char::from_u32_unchecked(*range.start()) };
+                let end = unsafe { char::from_u32_unchecked(*range.end()) };
+                if start == end {
```

Comment:
Issue: I don't actually think this is safe. I think there are public constructors that allow CPIV to get into this state. Please return a Serde serialization error instead.

Reply:
I think all constructors go through https://github.com/unicode-org/icu4x/blob/main/components/collections/src/codepointinvlist/utils.rs#L16, but we haven't actually declared that as an invariant anywhere.